drivers/video/video_arducam_mega.c

```diff
@@ -0,0 +1,1353 @@
+{
+	const struct arducam_mega_config *cfg = dev->config;
+	int ret = 0;
+
+	ret = arducam_mega_write_reg_wait(&cfg->bus, CAM_REG_CAPTURE_RESOLUTION, resolution, 10);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = arducam_mega_await_bus_idle(&cfg->bus, 10);
+	if (ret < 0) {
+		LOG_ERR("Bus idle wait failed after setting resolution");
+	}
+
+	return ret;
+}
+
+static int arducam_mega_check_connection(const struct device *dev)
+{
+	uint32_t cam_id;
+	const struct arducam_mega_config *cfg = dev->config;
+	struct arducam_mega_data *drv_data = dev->data;
+	int ret = 0;
+
+	ret = arducam_mega_await_bus_idle(&cfg->bus, 255);
+	if (ret < 0) {
+		LOG_ERR("Bus idle wait failed during connection check");
+		return ret;
+	}
+
+	ret = arducam_mega_read_reg(&cfg->bus, CAM_REG_SENSOR_ID, &cam_id);
+	if (ret < 0) {
+		LOG_ERR("Failed to read sensor ID");
+		return ret;
+	}
+	if (!(cam_id & 0x87)) {
+		LOG_ERR("arducam mega not detected, 0x%x\n", cam_id);
+		return -ENODEV;
+	}
+	drv_data->features = MEGA_HAS_DEFAULT;
+
+	switch (cam_id) {
+	case ARDUCAM_SENSOR_5MP_1:
+		fmts[8] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1944, VIDEO_PIX_FMT_RGB565);
+		fmts[17] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1944, VIDEO_PIX_FMT_JPEG);
+		fmts[26] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1944, VIDEO_PIX_FMT_YUYV);
+		drv_data->features |= MEGA_HAS_FOCUS | MEGA_HAS_COLORFX;
+		break;
+	case ARDUCAM_SENSOR_3MP_1:
+	case ARDUCAM_SENSOR_3MP_2:
+		fmts[8] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2048, 1536, VIDEO_PIX_FMT_RGB565);
+		fmts[17] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2048, 1536, VIDEO_PIX_FMT_JPEG);
+		fmts[26] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2048, 1536, VIDEO_PIX_FMT_YUYV);
+		drv_data->features |= MEGA_HAS_SHARPNESS | MEGA_HAS_COLORFX;
+		break;
+	case ARDUCAM_SENSOR_5MP_2:
+		fmts[8] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1936, VIDEO_PIX_FMT_RGB565);
+		fmts[17] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1936, VIDEO_PIX_FMT_JPEG);
+		fmts[26] = (struct video_format_cap)ARDUCAM_MEGA_VIDEO_FORMAT_CAP(
+			2592, 1936, VIDEO_PIX_FMT_YUYV);
+		drv_data->features |= MEGA_HAS_FOCUS | MEGA_HAS_COLORFX;
+		break;
+	default:
+		return -ENODEV;
+	}
+	drv_data->camera_id = cam_id;
+
+	return ret;
+}
+
+static int arducam_mega_set_fmt(const struct device *dev, struct video_format *fmt)
+{
+	struct arducam_mega_data *drv_data = dev->data;
+	int ret = 0;
+	int i = 0;
+
+	if (!memcmp(&drv_data->fmt, fmt, sizeof(drv_data->fmt))) {
+		/* nothing to do */
+		return 0;
+	}
+
+	ret = video_format_caps_index(fmts, fmt, &i);
+	if (ret < 0) {
+		LOG_ERR("Unsupported pixel format or resolution %s %ux%u",
+			VIDEO_FOURCC_TO_STR(fmt->pixelformat), fmt->width, fmt->height);
+		return ret;
+	}
+
```

Comment:
`arducam_mega_set_resolution()` is never called and the driver API is actually `arducam_mega_set_fmt()`.

Also, the video driver is expected to set the buffer size of the `fmt` struct in this function.

You can simply call `ret = arducam_mega_set_resolution(dev, i)` then copy-paste this code here:

https://github.com/zephyrproject-rtos/zephyr/blob/5741f3ee6be6283886e48577be5473de5f3c144d/drivers/video/video_stm32_dcmi.c#L214-L219